pkg/operator/observe_config.go

```diff
@@ -0,0 +1,167 @@
+	"k8s.io/client-go/util/flowcontrol"
+	"k8s.io/client-go/util/workqueue"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	operatorconfigclientv1alpha1 "github.com/openshift/cluster-kube-controller-manager-operator/pkg/generated/clientset/versioned/typed/kubecontrollermanager/v1alpha1"
+	operatorconfiginformerv1alpha1 "github.com/openshift/cluster-kube-controller-manager-operator/pkg/generated/informers/externalversions/kubecontrollermanager/v1alpha1"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"k8s.io/apimachinery/pkg/util/diff"
+)
+
+type observeConfigFunc func(kubernetes.Interface, *rest.Config, map[string]interface{}) (map[string]interface{}, error)
+
+type ConfigObserver struct {
+	operatorConfigClient operatorconfigclientv1alpha1.KubecontrollermanagerV1alpha1Interface
+
+	kubeClient   kubernetes.Interface
+	clientConfig *rest.Config
+
+	// queue only ever has one item, but it has nice error handling backoff/retry semantics
+	queue workqueue.RateLimitingInterface
+
+	rateLimiter flowcontrol.RateLimiter
+	observers   []observeConfigFunc
+}
+
+func NewConfigObserver(
+	operatorConfigInformer operatorconfiginformerv1alpha1.KubeControllerManagerOperatorConfigInformer,
+	kubeInformersForOpenshiftServiceCertSignerNamespace informers.SharedInformerFactory,
+	operatorConfigClient operatorconfigclientv1alpha1.KubecontrollermanagerV1alpha1Interface,
+	kubeClient kubernetes.Interface,
+	clientConfig *rest.Config,
+) *ConfigObserver {
+	c := &ConfigObserver{
+		operatorConfigClient: operatorConfigClient,
+		kubeClient:           kubeClient,
+		clientConfig:         clientConfig,
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ConfigObserver"),
+
+		rateLimiter: flowcontrol.NewTokenBucketRateLimiter(0.05 /*3 per minute*/, 4),
+		observers:   []observeConfigFunc{observeCABundleConfig},
+	}
+
+	operatorConfigInformer.Informer().AddEventHandler(c.eventHandler())
+	kubeInformersForOpenshiftServiceCertSignerNamespace.Core().V1().ConfigMaps().Informer().AddEventHandler(c.eventHandler())
+
+	return c
+}
+
+// sync reacts to a change in controller manager images.
+func (c ConfigObserver) sync() error {
+	var err error
+	observedConfig := map[string]interface{}{}
+
+	for _, observer := range c.observers {
+		observedConfig, err = observer(c.kubeClient, c.clientConfig, observedConfig)
+		if err != nil {
+			return err
+		}
+	}
+
+	operatorConfig, err := c.operatorConfigClient.KubeControllerManagerOperatorConfigs().Get("instance", metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	// don't worry about errors
+	currentConfig := map[string]interface{}{}
+	json.NewDecoder(bytes.NewBuffer(operatorConfig.Spec.ObservedConfig.Raw)).Decode(&currentConfig)
```

Comment:
@deads2k `operatorConfig` does not have a `.Spec.ObservedConfig` field here

Reply:
> @deads2k `operatorConfig` does not have a `.Spec.ObservedConfig` field here

ok, treat nil and empty the same.